ts/async_service.py

```diff
@@ -0,0 +1,156 @@
+    except PredictionException as e:
+        logger.error("Prediction error", exc_info=True)
+        return create_predict_response(None, req_id_map, e.message, e.error_code)
+    except Exception as ex:  # pylint: disable=broad-except
+        if "CUDA" in str(ex):
+            # Handles Case A: CUDA error: CUBLAS_STATUS_NOT_INITIALIZED (Close to OOM) &
+            # Case B: CUDA out of memory (OOM)
+            logger.error("CUDA out of memory", exc_info=True)
+            return create_predict_response(None, req_id_map, "Out of resources", 507)
+        else:
+            logger.warning("Invoking custom service failed.", exc_info=True)
+            return create_predict_response(None, req_id_map, "Prediction failed", 503)
+
+    if not isinstance(ret, list):
+        logger.warning(
+            "model: %s, Invalid return type: %s.",
+            context.model_name,
+            type(ret),
+        )
+        return create_predict_response(
+            None, req_id_map, "Invalid model predict output", 503
+        )
+
+    if len(ret) != len(input_batch):
+        logger.warning(
+            "model: %s, number of batch response mismatched, expect: %d, got: %d.",
+            context.model_name,
+            len(input_batch),
+            len(ret),
+        )
+        return create_predict_response(
+            None, req_id_map, "number of batch response mismatched", 503
+        )
+
+    duration = round((time.time() - start_time) * 1000, 2)
+    metrics.add_time(PREDICTION_METRIC, duration)
+
+    return create_predict_response(
+        ret, req_id_map, "Prediction success", 200, context=context
+    )
+
+
+class AsyncService(object):
+    def __init__(self, service):
+        self.service = service
+        self.service.predict = types.MethodType(predict, self.service)
+        self.in_queue = Queue()
+        self.out_queue = None
+        self.loop = None
+
+    def receive_requests(self):
+        while True:
+            logging.debug("Waiting for new message")
+            cmd, msg = retrieve_msg(self.service.cl_socket)
+
+            if cmd == b"I":
+                logging.debug(f"Putting msg in queue: {msg}")
+                self.in_queue.put(msg)
+            else:
+                logging.debug(f"Unexpected request: {cmd}")
+
+    async def call_predict(self, batch):
+        response = await self.service.predict(batch)
+        await self.out_queue.put(response)
+
+    def fetch_batches(self):
+        MAX_WAIT = 0.1
+        BATCH_SIZE = 1
+        while True:
+            st = time.time()
+            batch = []
+            try:
+                request = self.in_queue.get()
+                batch += request
+                while len(batch) < BATCH_SIZE and (time.time() - st) < MAX_WAIT:
+                    timeout = max(0, MAX_WAIT - (time.time() - st))
+                    request = self.in_queue.get(timeout=timeout)
+                    batch += request
```

Comment:
Nit: We may need to check here if we actually got a `request` or we timed out before adding to `batch`

Reply:
When .get() times out which is the only alternative to getting a request an Empty expection is thrown which is caught in line 126 and will just restart the batch window.